PanelLibrary: Adds get and getAll to the api

hugohaggmark opened the pull request:
**What this PR does / why we need it**:
This PR adds `GET /api/library-panels/:panelId` and `GET /api/library-panels/api` for Library/Shared/Reusable Panels.

I also added `Body()` method to the `NormalResponse` so I could verify the actual responses.

Finally but not least enjoy the review 🎉 

**Which issue(s) this PR fixes**:
Relates #29610

**Special notes for your reviewer**:
I did some refactoring of the tests because it was hard to debug a single test with the previous `libraryPanelScenario`. Please let me know if I went to far with that refactor :)

hugohaggmark committed:
PanelLibrary: Adds get to the API

aknuds1 committed:
Move middleware context handler logic to service (#29605)

* middleware: Move context handler to own service

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Co-authored-by: Emil Tullsted <sakjur@users.noreply.github.com>
Co-authored-by: Will Browne <wbrowne@users.noreply.github.com>

aknuds1 committed:
Chore: Configure go-ruleguard via golangci-lint (#28419)

* Chore: Configure go-ruleguard via golangci-lint

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

wbrowne committed:
Plugins: Add support for signature manifest V2 (#29240)

* add support for signing manifest v2

* add log and fix var name

* shorten comment

* improve comment

* remove unnecessary param

* improve naming

* reformat

* rename var

* refactor test

* remove unnecessary assert

* simplify test requirements

* add more test cases

* address feedback

* revert naming

* flip tracking missing

* fix check

* Trigger Build

kylebrandt committed:
Expressions: Field names from refId (#29755)

so one can see which expression data came from when visualizing

hugohaggmark committed:
Refactor: adds tests for get and getAll and cleans up other tests

torkelo committed:
GraphNG: Add test dashboard for null & and gaps rendering (#29769)

hugohaggmark committed:
Refactor: changed name on DTO

peterholmberg committed:
Alerting: Next gen Alerting page (#28397)

* create page and sidebar entry

* add components for query editor and definition

* split pane things

* add reducer and action

* implement split pane and update ui actions

* making things pretty

* Unify toolbar

* minor tweak to title prefix and some padding

* can create definitions

* fix default state

* add notificaion channel

* add wrappers to get correct spacing between panes

* include or exclude description

* implement query editor

* start on query result component

* update from master

* some cleanup and remove expressions touch ups

Co-authored-by: Torkel Ödegaard <torkel@grafana.com>

peterholmberg committed:
Expressions: Add placeholders to hint on input (#29773)

* math placeholder

* resample changes

aocenas committed:
Explore: Fix remounting of query row (#29771)

* Use passed data source in init

* Dummy commit

aocenas committed:
@grafana-runtime: Throw error if health check fails in DataSourceWithBackend (#29743)

mousedownmike committed:
Added httpMethod to webhook (#29780)

The values appear like they can be "POST" or "PUT" but there doesn't appear to be a place for that info in the documentation.

leeoniya committed:
GraphNG: update uPlot v1.5.0 (#29763)

leeoniya committed:
GraphNG: uPlot 1.5.1 (#29789)

haeringer committed:
CloudWatch: Add support for AWS DirectConnect ConnectionErrorCount metric (#29583)

ceh committed:
Chore: Rewrite opentsdb test to standard library (#29792)

aknuds1 committed:
Style guide: Document avoidance of globals in Go code (#29803)

* Chore: Document avoidance of globals in Go code

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aocenas committed:
Grafana-ui: Fix collapsible children sizing (#29776)

torkelo committed:
Sliders: Update behavior and style tweak (#29795)

* Sliders: Update behavior and style tweak

* More style tweaks, and changed new graph opacity option from 0 - 1 to 1 to 100

* Updated point size max

* Fixed hooks useCallback dependency issue

* Update test

Clarity-89 committed:
Grafana-UI: Support optgroup for MultiSelect (#29805)

* Grafana-UI: Support optgroup for MultiSelect

* Update packages/grafana-ui/src/components/Select/Select.story.tsx

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

jlosito committed:
Allow dependabot to check actions (#28159)

aknuds1 committed:
Chore: Use Header.Set method instead of Header.Add (#29804)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

Commit:
Elasticsearch: ensure query model has timeField configured in datasource settings (#29807)

dprokop committed:
GraphNG: Load uPlot path builders lazily (#29813)

aknuds1 committed:
Chore: Enable more go-ruleguard rules (#29781)

* Chore: Enable more go-ruleguard rules

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

oddlittlebird committed:
Docs: Update types-options.md (#29777)

kylebrandt committed:
Expressions: support ${my var} syntax (#29819)

ryantxu committed:
Chore: remove unused datasource status enum (#29827)

ryantxu committed:
QueryEditor: do not auto refresh on every update (#29762)

achatterjee-grafana committed:
Update style guide with correct usage of MS SQL (#29829)

* Added correct usage of MS SQL

* Updated as per Diana's comment.

* Fixed wording.

* Update contribute/style-guides/documentation-style-guide.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

aknuds1 committed:
Chore: Disable default golangci-lint filter (#29751)

* Disable default golangci-lint filter

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Chore: Fix linter warnings

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

torkelo commented on pkg/services/librarypanels/api.go:
```diff
@@ -49,3 +51,28 @@ func (lps *LibraryPanelService) deleteHandler(c *models.ReqContext) api.Response
+}
+
+// getAllHandler handles GET /api/library-panels/
+func (lps *LibraryPanelService) getAllHandler(c *models.ReqContext) api.Response {
```
think this will need some search, page  & page size at some point, but good to start simple :)

hugohaggmark replied:
Yes, keeping it simple for now :)

torkelo reviewed (approved):
Looks good.

One big thing that I forgot to think about for the db model & api is database numeric id vs uid (string guid). 

To support provisioning and referring to panel libraries in dashboard json where you do not know the DB id we need a uid for library panels. And think this is what panels in json (and in the HTTP api) should use. 

Ideas for persisted dashboard json (ie not runtime in DashboardModel) 
```typescript
panels: [
 {
    id: 10,    
    libraryPanel: {
      uid: 'asd12eadd12',
      name: 'Awesome graph',
    }
]
```

Something like that could be useful (to have uid & name), so when viewing a persisted json (saved to disk / git), you see the uid and the name (the name would be there just for readability). 

Switch to uid can be done in a separate PR. 

@aknuds1 do you have any feedback?

torkelo committed:
Sparkline: Fixes issue with sparkline that sent in custom fillColor instead of fillOpacity (#29825)

* Sparkline: Fixes issue with sparkline that sent in custom fillColor instead of fillOpacity

* Updated

hugohaggmark committed:
DashboardDS: Fixes display of long queries (#29808)

ivanahuckova committed:
Explore: Improve Explore performance but removing unnecessary re-renders (#29752)

* Fix Explore and Logs rerenders

* Create ReturnToDashboardButton

* Add test for new component

* Remove queryKeys change on keystroke

* Move rendering conditions to Button

* Remove redundant logic

dprokop committed:
GraphNG: Enable scale distribution configuration (#29684)

* Enable scale distribution configuration

* use uPlot's rangeLog() instead of rangeNum() for log scales

* merge master

* Remove unsupported log bases

Co-authored-by: Leon Sorokin <leeoniya@gmail.com>
Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

dprokop committed:
GraphNG: Context menu (#29745)

* Refactor Context menu and add Menu component to grafana/ui

* ContextMenuPlugin WIP

* Fix docs issues

* Remove Add annotations menu item from graph context menu

* ts ifx

hugohaggmark commented:
> Looks good.
> 
> One big thing that I forgot to think about for the db model & api is database numeric id vs uid (string guid).
> 
> To support provisioning and referring to panel libraries in dashboard json where you do not know the DB id we need a uid for library panels. And think this is what panels in json (and in the HTTP api) should use.
> 
> Ideas for persisted dashboard json (ie not runtime in DashboardModel)
> 
> ```ts
> panels: [
>  {
>     id: 10,    
>     libraryPanel: {
>       uid: 'asd12eadd12',
>       name: 'Awesome graph',
>     }
> ]
> ```
> 
> Something like that could be useful (to have uid & name), so when viewing a persisted json (saved to disk / git), you see the uid and the name (the name would be there just for readability).
> 
> Switch to uid can be done in a separate PR.
> 
> @aknuds1 do you have any feedback?

Good catch on the uid & name, will introduce that change after this PR.

ivanahuckova committed:
Pass row (#29839)

ivanahuckova committed:
Chore: Remove unused Loki and Cloudwatch syntax providers  (#29686)

* Remove syntax provider in Cloudwatch

* Remove syntax provider and refactor useLokiLabels

Clarity-89 committed:
Chore: Automatically infer types for dashgrid connected components  (#29818)

* Refactor DashboardPanel

* Refactor PanelEditor

* Fix missing initializer

* Update public/app/features/dashboard/components/PanelEditor/PanelEditor.tsx

Co-authored-by: Torkel Ödegaard <torkel@grafana.org>

* Simplify type

* Remove unused type

* Move prop connectors on top

Co-authored-by: Torkel Ödegaard <torkel@grafana.org>

hugohaggmark committed:
Select: Changes default menu placement for Select from auto to bottom (#29837)

* Select: Changes default menuPlacement from auto to bottom

* Tests: Fixes snapshot

dprokop committed:
Link suppliers: getLinks API update (#29757)

* ContextMenuPlugin WIP

* Remove Add annotations menu item from graph context menu

* ts ifx

* WIP

* Tests updates

* ts check fix

* Fix rebase

* Use replace function in angular graph data links

ivanahuckova committed:
Explore/Logs: Correctly display newlines in detected fields (#29541)

* Fix using of newlines in detected fields

* Wrapping depends on main selection

* Update props to be optional

* Add missing wrapLogMessage prop to test

hugohaggmark committed:
Varibles: Fixes so clicking on Selected will not include All (#29844)

ivanahuckova committed:
Loki: Remove unnecessary deduplication (#29421)

* Remove unnecessary deduplication

* Remove dedup test as we are not doing dedup on all logs  anymore

* Create unique ids in Loki

* Fix comment

* Fix comment

* Store prev response uids in usedUids

* Revert "Store prev response uids in usedUids"

This reverts commit 05c496e2a8150573513f2574cfef1407de96a72b.

* Add comment

Clarity-89 committed:
Docs: Add info about typing of connected props to Redux style guide (#29842)

* Add info about typing of connected props

* Update with suggestions

caetra committed:
Docs: Fix broken link in logs-panel (#29833)

Fixes #29799

Clarity-89 committed:
Display palette and colors for dark and light themes in storybook (#29848)

oddlittlebird committed:
MS SQL: Fix MS SQL add data source UI issues (#29832)

* Update config.html

* Update mssql.md

Co-authored-by: Torkel Ödegaard <torkel@grafana.com>

aknuds1 committed:
Backend: Remove more globals (#29644)

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

hybby committed:
add an example cloudwatch resource_arns() query that uses multiple tags (ref: #29499) (#29838)

torkelo committed:
GraphNG: Do not set fillColor from GraphNG only opacity (#29851)

* GraphNG: Do not set fillColor from GraphNG only opacity

* use color settings if they exist

Co-authored-by: Ryan McKinley <ryantxu@gmail.com>

ryantxu committed:
GraphNG: only initialize path builders once (#29863)

hugohaggmark committed:
Variables: Fixes so numerical sortorder works for options with null values (#29846)

leeoniya committed:
GraphNG: uPlot 1.5.2, dynamic stroke/fill, Flot-style hover points (#29866)

* upgrade uPlot 1.5.2 (dynamic stroke & fill, cursor.points API)

* flot-style hover points

* fix tests?

ryantxu committed:
QueryOptions:  Open QueryEditors: run queries after changing group options #29864

hugohaggmark commented:
@aknuds1 any feedback before I merge this?

hugohaggmark commented:
@wbrowne any feedback before I merge this?

hugohaggmark committed:
MSSQL: Config UI touches (#29834)

aknuds1 commented:
Checking.

aknuds1 commented on pkg/services/librarypanels/api.go:
```diff
@@ -17,13 +17,15 @@ func (lps *LibraryPanelService) registerAPIEndpoints() {
 
-// createHandler handles POST /api/library-panels.
-func (lps *LibraryPanelService) createHandler(c *models.ReqContext, cmd addLibraryPanelCommand) api.Response {
+// createHandler handles POST /api/library-panels
```
Massive nit, but I would keep the dot, since it's [standard](https://github.com/golang/go/wiki/CodeReviewComments#comment-sentences) for Go documentation comments :)
```suggestion
// createHandler handles POST /api/library-panels.
```

aknuds1 commented on pkg/services/librarypanels/api.go:
```diff
@@ -49,3 +51,28 @@ func (lps *LibraryPanelService) deleteHandler(c *models.ReqContext) api.Response
+
+// getHandler handles GET /api/library-panels/:panelId
+func (lps *LibraryPanelService) getHandler(c *models.ReqContext) api.Response {
+	libraryPanel, err := lps.getLibraryPanel(c, c.ParamsInt64(":panelId"))
+
+	if err != nil {
```
I would recommend not having a blank line between an error check and its origin, since it makes verifying harder for someone reading the code (as the connection is no longer obvious):
```suggestion
	libraryPanel, err := lps.getLibraryPanel(c, c.ParamsInt64(":panelId"))
	if err != nil {
```

hugohaggmark replied:
Sure I agree, will fix this in my next commit.

hugohaggmark replied:
Fixed

aknuds1 commented on pkg/services/librarypanels/api.go:
```diff
@@ -49,3 +51,28 @@ func (lps *LibraryPanelService) deleteHandler(c *models.ReqContext) api.Response
+
+// getAllHandler handles GET /api/library-panels/
+func (lps *LibraryPanelService) getAllHandler(c *models.ReqContext) api.Response {
+	libraryPanels, err := lps.getAllLibraryPanels(c)
+
+	if err != nil {
```
I would recommend not having a blank line between an error check and its origin, since it makes verifying harder for someone reading the code (as the connection is no longer obvious):
```suggestion
	libraryPanels, err := lps.getAllLibraryPanels(c)
	if err != nil {
```

hugohaggmark replied:
Sure I agree, will fix this in my next commit.

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
 	return err
 }
+
+// getLibraryPanels gets Library Panels based on statement and args
```
Why "Library Panels", if it's not referring to the LibraryPanel type? I'd think that if you're not referring to the specific type, you'd just write library panels?

hugohaggmark replied:
Fixed

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
+
+// getLibraryPanels gets Library Panels based on statement and args
+func getLibraryPanels(session *sqlstore.DBSession, statement interface{}, args ...interface{}) ([]LibraryPanel, error) {
+	libraryPanels := make([]LibraryPanel, 0)
+
+	err := session.SQL(statement, args...).Find(&libraryPanels)
```
I don't see the reason for so much whitespace :)
```suggestion
	libraryPanels := make([]LibraryPanel, 0)
	err := session.SQL(statement, args...).Find(&libraryPanels)
```

hugohaggmark replied:
Sure I agree, will fix this in my next commit.

hugohaggmark replied:
Removed the whole function

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
+
+// getLibraryPanel gets a Library Panel
+func (lps *LibraryPanelService) getLibraryPanel(c *models.ReqContext, panelID int64) (LibraryPanel, error) {
+	orgID := c.SignedInUser.OrgId
+	libraryPanel := LibraryPanel{}
+
```
I personally prefer removing this blank line, since these variables are specifically set by the closure.
```suggestion
	orgID := c.SignedInUser.OrgId
	libraryPanel := LibraryPanel{}
```

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
+	libraryPanel := LibraryPanel{}
+
+	err := lps.SQLStore.WithDbSession(context.Background(), func(session *sqlstore.DBSession) error {
+		libraryPanels, err := getLibraryPanels(session, "SELECT * FROM library_panel WHERE id=? and org_id=?", panelID, orgID)
+
+		if err != nil {
```
Can we just inline getLibraryPanels? That abstraction is so thin, I think it only makes it more difficult to read the code and see what actually happens. I'd find the following much easier to read:
```suggestion
		var libraryPanels []LibraryPanel
		err := session.SQL("SELECT * FROM library_panel WHERE id=? and org_id=?", panelID, orgID).Find(&libraryPanels)
		if err != nil {
```

hugohaggmark replied:
Sure, I can see your point. Will fix this in my next commit.

hugohaggmark replied:
Removed

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
+		if err != nil {
+			return err
+		} else if len(libraryPanels) != 1 {
+			return errLibraryPanelNotFound
```
Shouldn't this be treated as a bug instead?? It should be logged as an error and returned as 500 to the client, or maybe even ` panic` would be warranted since it would be a hard error on our side.

torkelo replied:
why would it be a bug? the panel is just not found with the specified id, would be a simple 404 in the API

aknuds1 replied:
@torkelo Sorry I didn't think about that case, I thought about the case where `len(libraryPanels) > 1`. I always take that case into account when I write these queries.

hugohaggmark replied:
That doesn't seem so YAGNI to me @aknuds1 ;) But sure I can add that as well in my next commit.

hugohaggmark replied:
Not YAGNI but added this as suggested.

aknuds1 replied:
That's the first time I hear of any connection between YAGNI and handling possible error cases. Handling possible error cases is only good code hygiene.

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
+// getAllLibraryPanels gets all Library Panels
+func (lps *LibraryPanelService) getAllLibraryPanels(c *models.ReqContext) ([]LibraryPanel, error) {
+	orgID := c.SignedInUser.OrgId
+	libraryPanels := make([]LibraryPanel, 0)
+
+	err := lps.SQLStore.WithDbSession(context.Background(), func(session *sqlstore.DBSession) error {
+		panels, err := getLibraryPanels(session, "SELECT * FROM library_panel WHERE org_id=?", orgID)
+
+		if err != nil {
```
Same concern as before, I don't think the `getLibraryPanels` abstraction should be used, inlining is more readable:
```suggestion
	var libraryPanels []LibraryPanel
	err := lps.SQLStore.WithDbSession(context.Background(), func(session *sqlstore.DBSession) error {
		err := session.SQL("SELECT * FROM library_panel WHERE org_id=?", orgID).Find(&libraryPanels)
		if err != nil {
```

hugohaggmark replied:
Sure, I can see your point. Will fix this in my next commit.

hugohaggmark replied:
Removed

aknuds1 commented on pkg/services/librarypanels/database.go:
```diff
@@ -62,3 +62,54 @@ func (lps *LibraryPanelService) deleteLibraryPanel(c *models.ReqContext, panelID
+			return err
+		}
+
+		libraryPanels = panels
+
+		return nil
```
```suggestion
		return nil
```

hugohaggmark replied:
Don't understand this comment, so did not change.

aknuds1 commented on pkg/services/librarypanels/librarypanels_test.go:
```diff
@@ -15,78 +16,251 @@ import (
-	libraryPanelScenario(t, "When an admin tries to create a library panel", "then it should fail if library panel already exists", func(t *testing.T) {
-		lps, context := setupTestEnv(t, models.ROLE_ADMIN, map[string]string{})
-		command := getCreateCommand(1, "Text - Library Panel")
+	t.Run("When an admin tries to create a library panel that already exists", func(t *testing.T) {
```
Why don't you use the `libraryPanelScenario` abstraction any longer, to define test cases? I think it makes sense, and is more ergonomic than the new pattern here? 
```suggestion
	t.Run("When an admin tries to create a library panel that already exists, it should fail", func(t *testing.T) {
```

hugohaggmark replied:
The simple answer is that I couldn't debug the individual t.Run within the libraryPanelScenario, but as you mentioned that I shouldn't have t.Run within t.Run I'll see what I can come up with.

hugohaggmark replied:
Fixed

aknuds1 commented on pkg/services/librarypanels/librarypanels_test.go:
```diff
@@ -15,78 +16,251 @@ import (
-		lps, context := setupTestEnv(t, models.ROLE_ADMIN, map[string]string{})
-		command := getCreateCommand(1, "Text - Library Panel")
+	t.Run("When an admin tries to create a library panel that already exists", func(t *testing.T) {
+		t.Run("then it should fail", func(t *testing.T) {
```
You should only use `t.Run` to define (sub) test cases, not to group assertions.
```suggestion
```

hugohaggmark replied:
Okidoki

hugohaggmark replied:
Fixed

aknuds1 commented on pkg/services/librarypanels/librarypanels_test.go:
```diff
@@ -15,78 +16,251 @@ import (
-		command := getCreateCommand(1, "Text - Library Panel")
+	t.Run("When an admin tries to create a library panel that already exists", func(t *testing.T) {
+		t.Run("then it should fail", func(t *testing.T) {
+			testScenario(t, getTestContext(), func(t *testing.T, ctx *testContext) {
```
I think that instead of this `getTextContext()` abstraction, it would be more Go style to provide a config object, with eventual parameters diverging from the defaults.

hugohaggmark replied:
Will look into this in my next commit as well.

hugohaggmark replied:
Not sure I understand this comment. Testscenario receives a pointer to a struct not a function. I invoked the function because I didn't see the point of declaring a context variable. Now I have declared an extra variable for each test instead.

aknuds1 reviewed (changes requested):
Please see my comments. I think mainly the tests should not be written with `t.Run` to group assertions - `t.Run` is meant for defining (sub) test cases, not grouping assertions. It's only confusing when running tests, if t.Run is used like this, since every invocation creates a sub-test that you can run in isolation and it becomes really noisy if every sub-test isn't actually a case.

MaximKipot committed:
Dashboard: Increase folder name size in search dashboard (#29821)

* Increase folder name size in search dashboard

* Use new fontsize in dashboard list panel as well

Co-authored-by: Torkel Ödegaard <torkel@grafana.com>

Clarity-89 committed:
Move colors demo under theme colors (#29873)

raymondchen625 committed:
Fix the broken link of XORM documentation (#29865)

vtorosyan committed:
Usage Stats: Allow to add additional metrics to the stats (#29774)

The change allows to add external, additional metrics which will be reported with the rest of hardcoded metrics, allowing to add metrics without a need to change the service itself.

hugohaggmark committed:
Update pkg/services/librarypanels/api.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

hugohaggmark committed:
Update pkg/services/librarypanels/database.go

Co-authored-by: Arve Knudsen <arve.knudsen@gmail.com>

torkelo committed:
InlineForms: Changes to make inline forms more flexible for query editors  (#29782)

* Chore: Remove internal import for QueryEditorRow

* WIP inline forms fixes

* Added segment group component

* Added component status doc tag

* Chore: add as prop to InlineLabel to control rendered element (#29876)

* Use div for InlineLabel & fix iconbuttons positioning

Co-authored-by: Elfo404 <gio.ricci@grafana.com>

ryantxu committed:
DataFrame: add a writable flag to fields (#29869)

vtorosyan committed:
Usage Stats: Introduce an interface for usage stats service (#29882)

Adding an interface type for usage stats service allows us to not depend on the implementation outside of the package, for example when testing we can easily mock the service

mariomann committed:
Update configure-docker.md (#29883)

It's not possible to stop an image, you can only start/stop container. I know it's just a little mistake and it sounds a bit like crumb cracker but your description is not correct.

wbrowne committed:
Alerting: Copy frame field labels to time series tags (#29886)

* Alerting: Copy frame field labels to time series tags

* keep comment

* add test case

ivanahuckova committed:
Keep query keys consistent (#29855)

aknuds1 committed:
Backend style guide: Add more guidelines (#29871)

* backend style guide: Add more guidelines

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

* Apply suggestions from code review

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

Co-authored-by: Emil Tullstedt <emil.tullstedt@grafana.com>

ryantxu committed:
GraphNG: support fill gradient (#29765)

torkelo committed:
SharedQuery: Fixes shared query editor now showing queries (#29849)

* SharedQuery: Fixes shared query editor now showing queries

* fixed updating state when query change

hugohaggmark committed:
Chore: changes after PR comments

aknuds1 committed:
Simplify tests

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

hugohaggmark commented:
@aknuds1 I fixed most of your comments, do you have anything else that needs to be done? (Changes requested)

aknuds1 commented:
@hugohaggmark I'm in the middle of something more pressing, but I am coming up with a suggestion for changes.

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into arve/libpanel-get

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>

aknuds1 reviewed (changes requested):
Thanks for applying my suggestions, I think the tests should be written according to a good pattern we have in the middleware tests, where devise wrappers for `t.Run` to have common setup logic. I've implemented this pattern in my branch [arve/libpanel-get](https://github.com/grafana/grafana/tree/arve/libpanel-get). Can you please merge that branch into your PR branch?

Eventually I can just push those changes to your branch if you want.

hugohaggmark commented:
> [arve/libpanel-get](https://github.com/grafana/grafana/tree/arve/libpanel-get).

Sure that looks good, thank you!

hugohaggmark committed:
Merge branch 'master' into hugoh/libpanel-get

hugohaggmark commented:
@aknuds1 is this ready for approval?

aknuds1 committed:
Merge remote-tracking branch 'origin/master' into hugoh/libpanel-get

aknuds1 committed:
Make test descriptions more concrete

Signed-off-by: Arve Knudsen <arve.knudsen@gmail.com>
